crypto/s2n_hmac.c

```diff
@@ -294,32 +312,40 @@ int s2n_hmac_digest_two_compression_rounds(struct s2n_hmac_state *state, void *o
 
 int s2n_hmac_free(struct s2n_hmac_state *state)
 {
-    GUARD(s2n_hash_free(&state->inner));
-    GUARD(s2n_hash_free(&state->inner_just_key));
-    GUARD(s2n_hash_free(&state->outer));
-    GUARD(s2n_hash_free(&state->outer_just_key));
+    if (state) {
+        GUARD(s2n_hash_free(&state->inner));
+        GUARD(s2n_hash_free(&state->inner_just_key));
+        GUARD(s2n_hash_free(&state->outer));
+        GUARD(s2n_hash_free(&state->outer_just_key));
+    }
 
-    return 0;
+    return S2N_SUCCESS;
 }
 
 int s2n_hmac_reset(struct s2n_hmac_state *state)
 {
+    PRECONDITION_POSIX(s2n_hmac_state_validate(state));
+    ENSURE_POSIX(state->hash_block_size != 0, S2N_ERR_PRECONDITION_VIOLATION);
     GUARD(s2n_hash_copy(&state->inner, &state->inner_just_key));
 
     uint64_t bytes_in_hash;
     GUARD(s2n_hash_get_currently_in_hash_total(&state->inner, &bytes_in_hash));
+    bytes_in_hash %= state->hash_block_size;
+    ENSURE_POSIX(bytes_in_hash <= UINT32_MAX, S2N_ERR_INTEGER_OVERFLOW);
     /* The length of the key is not private, so don't need to do tricky math here */
     state->currently_in_hash_block = bytes_in_hash % state->hash_block_size;
```

Comment:
Are you doing bytes_in_hash % state->hash_blob_size twice here?

Reply:
I was, but I reverted this change. Good catch!